zoho: a deleted or stale folder id returns 401 "R008 Unauthorized access" and fails hard instead of being handled as directory-not-found

### Before you start

- [x] I have searched the [forum](https://forum.rclone.org/) and the existing issues for this problem.
- [x] I have tested with the latest beta or stable release and can still reproduce the problem.
- [x] This is a reproducible bug, not a usage question (questions belong on the forum).

### Associated forum post URL

_No response_

### What is the problem you are having with rclone?

Zoho WorkDrive's listing API returns `401 "R008 Unauthorized access"` (not a 404) for a
folder id that no longer resolves to a listable folder - because the folder was deleted, or
the id never existed. I confirmed this live with a minimal probe (create a folder -> list it
= 200; trash it -> list it = 401 R008; list a random never-existed id = 401 R008). A freshly
refreshed OAuth token still gets the same R008, so it is not a token/scope problem.

The `zoho` backend does not handle this. In `backend/zoho/zoho.go`, `shouldRetry` lets the
401 fall through (and retrying a dead folder is futile and can escalate to a `429 F7008`
per-folder rate-limit penalty), and `listAll` surfaces it unchanged as
`couldn't list files: HTTP error 401 ... R008` - a hard failure that callers (object
lookups, `Put`) do not recover from.

This is a real gap versus other backends. Google Drive lists a directory with a query
(`'<id>' in parents and trashed=false`), so a dead/stale cached id simply returns an empty
result and self-heals; Zoho's `GET /files/<id>/files` returns R008 and fails hard. Because
rclone's vfs integration tests share one `Fs` (hence one dircache) across the whole package
and delete/recreate directories between tests, a stale `path -> dead id` cache entry is
essentially guaranteed, so the vfs suite fails on zoho where it passes on Drive. This also
mirrors a real long-running `rclone mount`/vfs whose cached folder is deleted by another
client.

### Steps to reproduce

1. Live probe: create a folder, trash it, then list it again -> `401 R008 Unauthorized
   access` instead of a 404.
2. In the integration suite, against a Zoho remote:
   `go test ./vfs -remote TestZoho: -v -run 'TestDir|TestFileOpen'`
   A test that resolves a directory from the shared dircache to an id a previous test
   deleted hits R008 and fails with `couldn't list files: ... R008`.

Measured on current master (`2e2a50cc4`), TestZoho:, quiesced: that run gives
**21 PASS / 5 FAIL / 1 SKIP**, the 5 failures all being R008 hard-fails
(TestDirFileOpen, TestDirMetadataExtension, TestFileOpenRead, TestFileOpenWrite, TestFileOpen).

### Expected behaviour

A folder id that returns R008 should be treated as a missing directory
(`fs.ErrorDirNotFound`), not retried, and a stale cached parent should be flushed so a later
create re-resolves it - the same graceful self-heal Drive gets for free. With that fix the
same run gives **26 PASS / 0 FAIL / 1 SKIP** (R008 still occurs but is handled), and a full
`test_all -backends zoho` stays clean at **96 PASS / 31 SKIP / 0 FAIL** with zero F7008. I
have a branch implementing this and will open it as a PR referencing this issue.


### rclone version

```text
rclone v1.74.3
- os/version: Microsoft Windows 11 Pro 25H2 25H2 (64 bit)
- os/kernel: 10.0.26200.8655 (x86_64)
- os/type: windows
- os/arch: amd64
- go/version: go1.26.4
- go/linking: static
- go/tags: cmount
```

(This is the stock release on PATH. The bug also reproduces on current master `2e2a50cc4`,
which is where the fix branch is based; the affected `listAll`/`shouldRetry` code in
`backend/zoho/zoho.go` is unchanged since this release, so released versions are equally
affected.)

### Operating system

Windows

### Which cloud storage system are you using?

Zoho WorkDrive

### Output of `rclone config redacted`

```ini
[TestZoho]
type = zoho
client_id = XXX
client_secret = XXX
region = eu
token = XXX
root_folder_id = XXX
```

### The command you were trying to run

```shell
go test ./vfs -remote TestZoho: -v -run 'TestDir|TestFileOpen' --dump-headers
```

### A log from the command with the `-vv` flag

```shell
=== RUN   TestDirFileOpen
2026/07/07 01:45:03 DEBUG : GET /api/v1/files/7wxaoaa7d4b2d38224a949f677a5c8546ced3/files?page[limit]=1000&page[next]=0 HTTP/1.1
2026/07/07 01:45:03 DEBUG : HTTP/2.0 401 Unauthorized
    Content-Type: application/vnd.api+json;charset=UTF-8
    (no Retry-After and no Www-Authenticate header - a plain 401)
    body: {"errors":[{"id":"R008","title":"Unauthorized access"}]}
    run.go:303: Failed to put "dir/file1" to "zoho root 'rclone-test-pojofuf3doqu'": couldn't list files: HTTP error 401 (401 Unauthorized) returned body: "{\"errors\":[{\"id\":\"R008\",\"title\":\"Unauthorized access\"}]}"
--- FAIL: TestDirFileOpen (9.24s)
```

### Anything else?

R008 confirmed live for both a deleted folder id and a random never-existed id (Zoho uses it
in place of a 404). The R008 401 carries no `Retry-After` or `Www-Authenticate` header - a
plain 401. Related zoho work already merged: #9571 (rate-limit/throttling). 

_Issue description drafted with AI assistance, edited and reviewed manually._

--- a/backend/zoho/throttle_test.go
+++ b/backend/zoho/throttle_test.go
@@ -2,6 +2,7 @@ package zoho
 
 import (
 	"context"
+	"errors"
 	"net/http"
 	"sync"
 	"testing"
@@ -73,6 +74,16 @@ func TestShouldRetry(t *testing.T) {
 		assert.True(t, retry)
 	})
 
+	// A bare 401 R008 body means the folder was deleted (Zoho no longer accepts
+	// its id). It is unrecoverable, so it is NOT retried.
+	t.Run("401 R008 deleted folder is not retried", func(t *testing.T) {
+		f := newTestFs()
+		resp := &http.Response{StatusCode: 401}
+		err := errors.New(`HTTP error 401: {"errors":[{"id":"R008","title":"Unauthorized access"}]}`)
+		retry, _ := f.shouldRetry(ctx, resp, err)
+		assert.False(t, retry)
+	})
+
 	// A cancelled context is never retried.
 	t.Run("cancelled context aborts", func(t *testing.T) {
 		f := newTestFs()
@@ -81,6 +92,26 @@ func TestShouldRetry(t *testing.T) {
 		retry, _ := f.shouldRetry(cctx, nil, nil)
 		assert.False(t, retry)
 	})
+}
+
+// isMissingResourceErr must match only a genuine 401 R008 body, so an unrelated
+// error that merely mentions R008 (or an R008 body on some other status) is not
+// mistaken for a deleted resource.
+func TestIsMissingResourceErr(t *testing.T) {
+	r008 := errors.New(`HTTP error 401: {"errors":[{"id":"R008","title":"Unauthorized access"}]}`)
+
+	// A 401 R008 body is a missing resource.
+	assert.True(t, isMissingResourceErr(&http.Response{StatusCode: 401}, r008))
+
+	// An R008 body on a non-401 status is not (guards the ungated substring).
+	assert.False(t, isMissingResourceErr(&http.Response{StatusCode: 500}, r008))
+
+	// A 401 that is not R008 is not a missing resource.
+	assert.False(t, isMissingResourceErr(&http.Response{StatusCode: 401}, errors.New("HTTP error 401: expired_token")))
+
+	// No response or no error is not a missing resource.
+	assert.False(t, isMissingResourceErr(nil, r008))
+	assert.False(t, isMissingResourceErr(&http.Response{StatusCode: 401}, nil))
 }
 
 // TestThrottleEpisode covers the once-per-episode logging state machine that

--- a/backend/zoho/zoho.go
+++ b/backend/zoho/zoho.go
@@ -657,12 +657,20 @@ func (f *Fs) logThrottle(wait time.Duration, err error) {
 	}
 }
 
+// isMissingResourceErr reports whether resp/err is Zoho's 401 "R008 Unauthorized
+// access" - returned (not a 404) for a resource id (folder or file) that was
+// deleted or never existed. A freshly refreshed token still gets it, so it is a
+// missing resource, not a token problem.
+func isMissingResourceErr(resp *http.Response, err error) bool {
+	return resp != nil && resp.StatusCode == 401 && err != nil && strings.Contains(err.Error(), "R008")
+}
+
 // shouldRetry reports whether the given resp and err deserve to be retried.
 //
 // A 429 is honoured via the Retry-After header (falling back to 60s plus a
 // margin) and starts or continues a throttling episode; expired OAuth tokens
 // are retried, missing OAuth scopes abort, and standard HTTP retry conditions
-// are also handled.
+// are also handled. A missing folder (R008) is not retried.
 //
 // Returns whether to retry, and the err as a convenience.
 func (f *Fs) shouldRetry(ctx context.Context, resp *http.Response, err error) (bool, error) {
@@ -687,6 +695,11 @@ func (f *Fs) shouldRetry(ctx context.Context, resp *http.Response, err error) (b
 	if resp != nil && resp.StatusCode == 401 && len(resp.Header["Www-Authenticate"]) == 1 && strings.Contains(resp.Header["Www-Authenticate"][0], "expired_token") {
 		authRetry = true
 		fs.Debugf(nil, "Should retry: %v", err)
+	}
+
+	// A missing resource never reappears, and retrying escalates to a 429 F7008.
+	if isMissingResourceErr(resp, err) {
+		return false, err
 	}
 	if resp != nil && resp.StatusCode == 429 {
 		// Zoho's listing API is heavily rate limited and tells us how long to
@@ -766,6 +779,16 @@ func (f *Fs) readMetaDataForPath(ctx context.Context, path string) (info *api.It
 		return false
 	})
 	if err != nil {
+		if err == fs.ErrorDirNotFound {
+			// The cached parent directory id is stale: its folder was deleted, so
+			// listing it returned R008 (mapped to ErrorDirNotFound). Flush the stale
+			// entry so a later create re-resolves (and recreates) the parent, and
+			// report the object as not found - it cannot exist if its parent is gone.
+			parent := strings.TrimSuffix(path[:len(path)-len(leaf)], "/")
+			fs.Debugf(f, "readMetaDataForPath %q: parent %q stale (R008), flushing dircache", path, parent)
+			f.dirCache.FlushDir(parent)
+			return nil, fs.ErrorObjectNotFound
+		}
 		return nil, err
 	}
 	if !found {
@@ -994,7 +1017,13 @@ OUTER:
 			return f.shouldRetry(ctx, resp, err)
 		})
 		if err != nil {
-			return found, fmt.Errorf("couldn't list files: %w", err)
+			// Surface a missing folder as directory-not-found so dircache/the VFS
+			// treat it as gone instead of hard-failing on a stale id.
+			if isMissingResourceErr(resp, err) {
+				fs.Debugf(f, "listAll %q: R008 unauthorized - treating as directory not found", dirID)
+				return false, fs.ErrorDirNotFound
+			}
+			return false, fmt.Errorf("couldn't list files: %w", err)
 		}
 		if len(result.Items) == 0 {
 			break

